File: src/main/java/org/mininuniver/interactiveMap/services/FloorService.java
/*
 * This file is part of mininuniver-interactive-map-service.
 *
 * Copyright (C) 2025 Eiztrips
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package org.mininuniver.interactiveMap.services;

import jakarta.persistence.EntityNotFoundException;
import jakarta.persistence.OptimisticLockException;
import lombok.RequiredArgsConstructor;
import org.mininuniver.interactiveMap.dto.models.*;
import org.mininuniver.interactiveMap.dto.models.node.NodeDTO;
import org.mininuniver.interactiveMap.dto.models.path.PathDTO;
import org.mininuniver.interactiveMap.dto.models.room.RoomDTO;
import org.mininuniver.interactiveMap.dto.models.stairs.StairsDTO;
import org.mininuniver.interactiveMap.dto.models.floor.FloorShortDTO;
import org.mininuniver.interactiveMap.dto.models.floor.FloorDTO;
import org.mininuniver.interactiveMap.models.*;
import org.mininuniver.interactiveMap.repositories.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigInteger;
import java.util.*;

@Service
@RequiredArgsConstructor
public class FloorService {

    private final FloorRepository floorRepository;
    private final RoomRepository roomRepository;
    private final StairsRepository stairsRepository;
    private final NodeRepository nodeRepository;
    private final PathRepository pathRepository;

    public List<FloorShortDTO> getAllFloors() {
        List<Floor> floors = floorRepository.findAll();
        return floors.stream()
                .map(FloorShortDTO::new)
                .toList();
    }

    public MapDTO getMapData(int number) {
        FloorDTO floor = new FloorDTO(floorRepository.findByNumber(number)
                .orElseThrow(() -> new EntityNotFoundException("Этаж не найден")));

        List<RoomDTO> rooms = roomRepository.findByFloorId(floor.getId())
                .stream()
                .map(RoomDTO::new)
                .toList();

        List<StairsDTO> stairs = stairsRepository.findByFloorId(floor.getId())
                .stream()
                .map(StairsDTO::new)
                .toList();

        List<PathDTO> paths = pathRepository.findByFloorId(floor.getId())
                .stream()
                .map(PathDTO::new)
                .toList();

        List<NodeDTO> nodes = nodeRepository.findByFloorId(floor.getId())
                .stream()
                .map(NodeDTO::new)
                .toList();

        return new MapDTO(floor, rooms, stairs, nodes, paths);
    }

    @Transactional
    public MapDTO updateFloorData(Integer number, MapDTO mapDTO) {
        Floor floor = floorRepository.findByNumber(number).orElseGet(Floor::new);
        floor.setNumber(number);
        floor.setName(mapDTO.getFloor().getName());
        floor.setPoints(mapDTO.getFloor().getPoints());
        floor = floorRepository.save(floor);

        roomRepository.deleteAllByFloorId(floor.getId());
        nodeRepository.deleteAllByFloorId(floor.getId());
        stairsRepository.deleteAllByFloorId(floor.getId());

        Map<Long, Long> nodeIdMapping = new HashMap<>();

        for (NodeDTO node : mapDTO.getNodes()) {
            Long oldId = node.getId();
            node.setId(null);
            node.setFloorId(floor.getId());
            Node saved = nodeRepository.save(new Node(node));
            nodeIdMapping.put(oldId, saved.getId());
        }

        for (Map.Entry<Long, Long> entry : nodeIdMapping.entrySet()) {
            NodeDTO node = new NodeDTO(nodeRepository.findById(entry.getValue()).orElseThrow());

            Long[] oldNeighbors = node.getNeighbors();

            if (oldNeighbors != null) {
                Long[] newNeighbors = Arrays.stream(oldNeighbors)
                        .map(n -> nodeIdMapping.getOrDefault(n, n))
                        .toArray(Long[]::new);
                node.setNeighbors(newNeighbors);
                nodeRepository.save(new Node(node));
            }
        }

        for (RoomDTO room : mapDTO.getRooms()) {
            room.setId(null);
            room.setFloorId(floor.getId());
            if (room.getNodeId() != null) {
                Long oldNodeId = room.getNodeId();
                if (nodeIdMapping.containsKey(oldNodeId)) {
                    room.setNodeId(nodeIdMapping.get(oldNodeId));
                }
            }
            roomRepository.save(new Room(room));
        }

        for (StairsDTO stair : mapDTO.getStairs()) {
            stair.setFloorId(floor.getId());
            stair.setId(null);
            stair.setNodeId(nodeIdMapping.get(stair.getNodeId()));
            stairsRepository.save(new Stairs(stair));
        }

        return mapDTO;
    }

    @Transactional
    public void deleteFloor(int number) {
        Floor floor = floorRepository.findByNumber(number)
                .orElseThrow(() -> new EntityNotFoundException("Этаж не найден"));

        try {
            roomRepository.deleteAllByFloorId(floor.getId());
            stairsRepository.deleteAllByFloorId(floor.getId());
            nodeRepository.deleteAllByFloorId(floor.getId());
            floorRepository.delete(floor);
        } catch (OptimisticLockException e) {
            throw new RuntimeException("Ошибка при удалении этажа: " + e.getMessage());
        }
    }

}

File: src/main/java/org/mininuniver/interactivemap/services/FloorService.kt
/*
 * This file is part of mininuniver-interactive-map-service.
 *
 * Copyright (C) 2025 Eiztrips
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package org.mininuniver.interactivemap.services

import jakarta.persistence.EntityNotFoundException
import jakarta.persistence.OptimisticLockException
import org.mininuniver.interactivemap.dto.FloorDTO
import org.mininuniver.interactivemap.models.Floor
import org.mininuniver.interactivemap.repositories.EdgeRepository
import org.mininuniver.interactivemap.repositories.FloorRepository
import org.mininuniver.interactivemap.repositories.NodeRepository
import org.mininuniver.interactivemap.repositories.RoomRepository
import org.mininuniver.interactivemap.repositories.StairsRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class FloorService @Autowired constructor(
    private val floorRepository: FloorRepository,
    private val roomRepository: RoomRepository,
    private val edgeRepository: EdgeRepository,
    private val stairsRepository: StairsRepository,
    private val nodeRepository: NodeRepository
) {

    fun getFloorData(id: Int): FloorDTO {
        val floor = floorRepository.findById(id)
            .orElseThrow { RuntimeException("Этаж не найден") }

        val rooms = roomRepository.findByFloorId(floor.id)
        val edges = edgeRepository.findByFloorId(floor.id)
        val stairs = stairsRepository.findByFloorId(floor.id)
        val nodes = nodeRepository.findByFloorId(floor.id)

        return FloorDTO(floor, rooms, edges, stairs, nodes)
    }

    @Transactional
    fun saveOrUpdateFloorData(id: Int, floorDTO: FloorDTO): FloorDTO {
        var floor = floorRepository.findById(id).orElse(Floor())
        floor.floorNumber = id
        floor.name = floorDTO.floor.name
        floor.points = floorDTO.floor.points
        floor = floorRepository.save(floor)

        roomRepository.deleteAllByFloorId(floor.id)
        nodeRepository.deleteAllByFloorId(floor.id)
        edgeRepository.deleteAllByFloorId(floor.id)
        stairsRepository.deleteAllByFloorId(floor.id)

        val nodeIdMapping = HashMap<Int, Int>()

        for (node in floorDTO.nodes) {
            val oldId = node.id
            node.id = null
            node.floorId = floor.id
            node.nodeNumber = oldId
            val saved = nodeRepository.save(node)
            oldId?.let { nodeIdMapping[it] = saved.id!! }
        }

        for ((oldId, newId) in nodeIdMapping) {
            val node = nodeRepository.findById(newId).orElseThrow()
            val oldNeighbors = floorDTO.nodes.find { it.id == oldId }?.neighbors

            oldNeighbors?.let {
                val newNeighbors = it.map { n -> nodeIdMapping.getOrDefault(n, n) }.toIntArray()
                node.neighbors = newNeighbors
                nodeRepository.save(node)
            }
        }

        for (edge in floorDTO.edges) {
            edge.id = null
            edge.floorId = floor.id
            val newNodes = edge.nodes.map { n -> nodeIdMapping.getOrDefault(n, n) }.toIntArray()
            edge.nodes = newNodes
            edgeRepository.save(edge)
        }

        for (room in floorDTO.rooms) {
            room.id = null
            room.floorId = floor.id
            room.nodeId?.let { oldNodeId ->
                if (nodeIdMapping.containsKey(oldNodeId)) {
                    room.nodeId = nodeIdMapping[oldNodeId]
                }
            }
            roomRepository.save(room)
        }

        for (stair in floorDTO.stairs) {
            stair.id = null
            stair.floorId = floor.id
            stairsRepository.save(stair)
        }

        return floorDTO
    }

    @Transactional
    fun deleteFloor(number: Int) {
        val floor = floorRepository.findByFloorNumber(number)
            .orElseThrow { EntityNotFoundException("Этаж не найден") }

        try {
            roomRepository.deleteAllByFloorId(floor.id!!)
            edgeRepository.deleteAllByFloorId(floor.id!!)
            stairsRepository.deleteAllByFloorId(floor.id!!)
            nodeRepository.deleteAllByFloorId(floor.id!!)
            floorRepository.delete(floor)
        } catch (e: OptimisticLockException) {
            throw RuntimeException("Ошибка при удалении этажа: ${e.message}")
        }
    }
}